hi-ml-histopathology/testhisto/testhisto/datamodules/test_slides_datamodule.py

```diff
@@ -0,0 +1,107 @@
+    tmp_root_dir = tmpdir_factory.mktemp("mock_wsi")
+    wsi_generator = MockWSIGenerator(
+        tmp_path=tmp_root_dir,
+        mock_type=MockWSIType.PATHMNIST,
+        batch_size=1,
+        n_samples=4,
+        n_repeat_diag=N_REPEAT_DIAG,
+        n_repeat_tile=2,
+        n_channels=3,
+        n_levels=3,
+        tile_size=28,
+        background_val=255
+    )
+    wsi_generator.generate_mock_wsi()
+    return tmp_root_dir
+
+
+@pytest.mark.skipif(no_gpu, reason="Test requires GPU")
+@pytest.mark.gpu
+def test_tiling_on_the_fly(mock_wsi_root_dir: py.path.local) -> None:
+    batch_size, tile_count, tile_size, level, channels = 1, 16, 28, 0, 3
+    datamodule = MockSlidesDataModule(
+        root_path=mock_wsi_root_dir, batch_size=batch_size, tile_count=tile_count, tile_size=tile_size, level=level
+    )
+    dataloader = datamodule.train_dataloader()
+    for sample in dataloader:
+        # sanity check for expected shape
+        tiles, wsi_id = sample["image"], sample["slide_id"][0]
+        assert tiles.shape == (batch_size, tile_count, channels, tile_size, tile_size)
+
+        # check tiling on the fly
+        original_tile = np.load(str(mock_wsi_root_dir.join(f"{wsi_id}_tile.npy")))[0]
+        for i in range(tile_count):
+            assert (original_tile == tiles[0, i].numpy()).all()
+
+
+@pytest.mark.skipif(no_gpu, reason="Test requires GPU")
+@pytest.mark.gpu
+def test_tiling_without_fixed_tile_count(mock_wsi_root_dir: py.path.local) -> None:
+    batch_size, tile_count, tile_size, level = 1, None, 28, 0
+    min_expected_tile_count = 16
+    datamodule = MockSlidesDataModule(
+        root_path=mock_wsi_root_dir, batch_size=batch_size, tile_count=tile_count, tile_size=tile_size, level=level
+    )
+    dataloader = datamodule.train_dataloader()
+    for sample in dataloader:
+        tiles = sample["image"]
+        assert tiles.shape[1] >= min_expected_tile_count
+
+
+@pytest.mark.skipif(no_gpu, reason="Test requires GPU")
+@pytest.mark.gpu
+@pytest.mark.parametrize("level", [0, 1, 2])
+def test_multi_resolution_tiling(level: int, mock_wsi_root_dir: py.path.local) -> None:
+    batch_size, tile_count, channels = 1, 16, 3
+    tile_size = 28 // 2 ** level
+    datamodule = MockSlidesDataModule(
+        root_path=mock_wsi_root_dir, batch_size=batch_size, tile_count=tile_count, tile_size=tile_size, level=level
+    )
+    dataloader = datamodule.train_dataloader()
+    for sample in dataloader:
+        # sanity check for expected shape
+        tiles, wsi_id = sample["image"], sample["slide_id"][0]
+        assert tiles.shape == (batch_size, tile_count, channels, tile_size, tile_size)
+
+        # check tiling on the fly at different resolutions
+        original_tile = np.load(str(mock_wsi_root_dir.join(f"{wsi_id}_tile.npy")))[0]
+        for i in range(tile_count):
+            assert (original_tile[:, :: 2 ** level, :: 2 ** level] == tiles[0, i].numpy()).all()
```

Comment:
A comment here referring to the fact that multi resolution versions of the mock image has been created via 2 factor downsampling might help the reader